src/commp.rs

```diff
@@ -0,0 +1,261 @@
+            Ok(cs)
+        }
+    }
+}
+
+// logic partly copied from Lotus' PadReader which is also in go-fil-markets
+// figure out how big this piece will be when padded
+fn padded_size(size: u64) -> u64 {
+    let logv = 64 - size.leading_zeros();
+    let sect_size = (1 as u64) << logv;
+    let bound = u64::from(UnpaddedBytesAmount::from(SectorSize(sect_size)));
+    if size <= bound {
+        return bound;
+    }
+    return u64::from(UnpaddedBytesAmount::from(SectorSize(1 << (logv + 1))));
+}
+
+// reimplemented from rust-fil-proofs/src/storage_proofs/pieces.rs
+fn local_generate_piece_commitment_bytes_from_source<H: Hasher, R: Sized>(
+    source: &mut R,
+    padded_piece_size: usize,
+) -> anyhow::Result<Fr32Ary>
+where
+    R: io::Read,
+{
+    let mut buf = [0; NODE_SIZE];
+    let mut reader = BufReader::new(source);
+    let parts = (padded_piece_size as f64 / NODE_SIZE as f64).ceil() as usize;
+
+    info!("Calculating merkle ...");
+
+    let tree = MerkleTree::<H::Domain, H::Function>::try_from_iter((0..parts).map(|_| {
+        reader.read_exact(&mut buf)?;
+        <H::Domain as Domain>::try_from_bytes(&buf)
+    }))?;
+
+    let mut comm_p_bytes = [0; NODE_SIZE];
+    let comm_p = tree.root();
+    comm_p
+        .write_bytes(&mut comm_p_bytes)
+        .expect("borked at extracting commp bytes");
+
+    info!("CommP from merkle root: {:?}", comm_p_bytes);
+    Ok(comm_p_bytes)
+}
+
+// same as local_generate_piece_commitment_bytes_from_source but uses a custom MerkleTree
+// caching store from multistore.rs
+fn local_multistore_generate_piece_commitment_bytes_from_source<H: Hasher, R: Sized>(
+    source: &mut R,
+    padded_piece_size: usize,
+) -> anyhow::Result<Fr32Ary>
+where
+    R: io::Read,
+{
+    let mut buf = [0; NODE_SIZE];
+    let mut reader = BufReader::new(source);
+    let parts = (padded_piece_size as f64 / NODE_SIZE as f64).ceil() as usize;
+
+    info!("Calculating merkle with multistore ...");
+
+    let tree =
+        MerkleTreeMultiStore::<H::Domain, H::Function>::try_from_iter((0..parts).map(|_| {
+            reader.read_exact(&mut buf)?;
+            <H::Domain as Domain>::try_from_bytes(&buf)
+        }))?;
+
+    let mut comm_p_bytes = [0; NODE_SIZE];
+    let comm_p = tree.root();
+    comm_p
+        .write_bytes(&mut comm_p_bytes)
+        .expect("borked at extracting commp bytes");
+
+    info!("CommP from merkle root: {:?}", comm_p_bytes);
+    Ok(comm_p_bytes)
+}
+
+fn padded<R: Sized>(inp: &mut R, size: u64) -> PadReader<&mut R>
+where
+    R: io::Read,
+{
+    let padded_size = padded_size(size);
+
+    let pad_reader = PadReader {
+        size: usize::try_from(size).unwrap(),
+        padsize: usize::try_from(padded_size).unwrap(),
+        pos: 0,
+        inp: inp,
+    };
+
+    return pad_reader;
+}
+
+#[allow(dead_code)]
+pub fn generate_commp_filecoin_proofs<R: Sized>(
+    inp: &mut R,
+    size: u64,
+) -> Result<CommP, std::io::Error>
+where
+    R: io::Read,
```

Comment:
I think (not sure though) that you can write `R: io::Read + Sized` here and using just `generate_commp_filecoin_proofs<R>()`. This way you have all the bounds requirements in one place.